File: go/indicators/jeanphilippepoton/fractalbands/fractalbands.go
package fractalbands

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
	"zpano/indicators/core/outputs"
)

// FractalBands computes the Fractal Bands indicator.
//
// FRASMA2 center line with upper/lower bands scaled by alpha^H where H is
// the local Hurst exponent estimated from the Fractal Graph Dimension Index.
//
// The indicator is not primed during the first `period - 1` updates.
type FractalBands struct {
	mu sync.RWMutex
	core.LineIndicator
	barFunc      entities.BarFunc
	quoteFunc    entities.QuoteFunc
	tradeFunc    entities.TradeFunc
	window       []float64
	closes       []float64
	period       int
	periodMinus1 int
	normalSpeed  int
	alpha        float64
	windowCount  int
	primed       bool
	logDenom     float64
	ln2          float64
	invPeriodSq  float64
	frasma2      float64
	upperBand    float64
	lowerBand    float64
}

// NewFractalBands returns an instance of the indicator created using supplied parameters.
func NewFractalBands(p *Params) (*FractalBands, error) {
	const (
		invalid = "invalid fractal bands parameters"
		fmts    = "%s: %s"
		fmtw    = "%s: %w"
		fmtn    = "fban(%d,%d,%g%s)"
	)

	period := p.Period
	if period < 2 {
		return nil, fmt.Errorf(fmts, invalid, "period should be greater than 1")
	}

	normalSpeed := p.NormalSpeed
	if normalSpeed < 1 {
		return nil, fmt.Errorf(fmts, invalid, "normal_speed should be greater than 0")
	}

	alpha := p.Alpha
	if alpha <= 0.0 {
		return nil, fmt.Errorf(fmts, invalid, "alpha should be greater than 0")
	}

	bc := p.BarComponent
	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	qc := p.QuoteComponent
	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	tc := p.TradeComponent
	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	var (
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	mnemonic := fmt.Sprintf(fmtn, period, normalSpeed, alpha, core.ComponentTripleMnemonic(bc, qc, tc))
	desc := "Fractal bands " + mnemonic

	periodMinus1 := period - 1

	f := &FractalBands{
		barFunc:      barFunc,
		quoteFunc:    quoteFunc,
		tradeFunc:    tradeFunc,
		window:       make([]float64, period),
		closes:       make([]float64, 0, 256),
		period:       period,
		periodMinus1: periodMinus1,
		normalSpeed:  normalSpeed,
		alpha:        alpha,
		logDenom:     math.Log(2.0 * float64(periodMinus1)),
		ln2:          math.Log(2.0),
		invPeriodSq:  1.0 / float64(period*period),
		frasma2:      math.NaN(),
		upperBand:    math.NaN(),
		lowerBand:    math.NaN(),
	}

	f.LineIndicator = core.NewLineIndicator(mnemonic, desc, barFunc, quoteFunc, tradeFunc, f.Update)

	return f, nil
}

// IsPrimed indicates whether the indicator is primed.
func (f *FractalBands) IsPrimed() bool {
	f.mu.RLock()
	defer f.mu.RUnlock()

	return f.primed
}

// Metadata describes the output data of the indicator.
func (f *FractalBands) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.FractalBands,
		f.LineIndicator.Mnemonic,
		f.LineIndicator.Description,
		[]core.OutputText{
			{Mnemonic: f.LineIndicator.Mnemonic, Description: f.LineIndicator.Description},
			{Mnemonic: f.LineIndicator.Mnemonic + " upper", Description: f.LineIndicator.Description + " Upper Band"},
			{Mnemonic: f.LineIndicator.Mnemonic + " lower", Description: f.LineIndicator.Description + " Lower Band"},
			{Mnemonic: f.LineIndicator.Mnemonic + " band", Description: f.LineIndicator.Description + " Band"},
		},
	)
}

// Update updates the value of the indicator given the next sample.
// Returns the FRASMA2 value. Use UpdateAll() for all three outputs.
func (f *FractalBands) Update(sample float64) float64 {
	if math.IsNaN(sample) {
		return sample
	}

	f.mu.Lock()
	defer f.mu.Unlock()

	period := f.period
	periodMinus1 := f.periodMinus1

	// Accumulate close history for SMA computation.
	f.closes = append(f.closes, sample)

	// Fill the FGDI window.
	if f.windowCount < period {
		f.window[f.windowCount] = sample
		f.windowCount++

		if f.windowCount < period {
			return math.NaN()
		}

		f.primed = true
	} else {
		for i := 0; i < periodMinus1; i++ {
			f.window[i] = f.window[i+1]
		}

		f.window[periodMinus1] = sample
	}

	// Find min/max for normalization.
	priceMax := f.window[0]
	priceMin := f.window[0]

	for k := 1; k < period; k++ {
		if f.window[k] > priceMax {
			priceMax = f.window[k]
		}

		if f.window[k] < priceMin {
			priceMin = f.window[k]
		}
	}

	priceRange := priceMax - priceMin

	var fgdi float64

	if priceRange <= 0.0 {
		fgdi = 0.0
	} else {
		// Compute normalized path length: period points, period-1 segments.
		priorNorm := (f.window[0] - priceMin) / priceRange
		length := 0.0

		for k := 1; k < period; k++ {
			currNorm := (f.window[k] - priceMin) / priceRange
			diff := currNorm - priorNorm
			length += math.Sqrt(diff*diff + f.invPeriodSq)
			priorNorm = currNorm
		}

		if length > 0.0 {
			fgdi = 1.0 + (math.Log(length)+f.ln2)/f.logDenom
		} else {
			fgdi = 0.0
		}
	}

	// Hurst exponent.
	hurst := 2.0 - fgdi
	if hurst < 0.01 {
		hurst = 0.01
	}

	trailDim := 1.0 / hurst
	beta := trailDim / 2.0
	speed := int(math.Round(float64(f.normalSpeed) * beta))

	if speed < 1 {
		speed = 1
	}

	// FRASMA2: SMA of close over 'speed' bars ending at current position.
	nCloses := len(f.closes)
	if speed > nCloses {
		f.frasma2 = math.NaN()
		f.upperBand = math.NaN()
		f.lowerBand = math.NaN()

		return math.NaN()
	}

	smaSum := 0.0
	for k := nCloses - speed; k < nCloses; k++ {
		smaSum += f.closes[k]
	}

	frasma2Val := smaSum / float64(speed)

	// Deviation over the FGDI lookback window (period bars).
	sqSum := 0.0
	for k := 0; k < period; k++ {
		res := f.window[k] - frasma2Val
		sqSum += res * res
	}

	deviation := 2.0 * math.Sqrt(sqSum/float64(period))

	// Fractal bands.
	bandMult := deviation * math.Pow(f.alpha, hurst)
	upperBand := frasma2Val + bandMult
	lowerBand := frasma2Val - bandMult

	f.frasma2 = frasma2Val
	f.upperBand = upperBand
	f.lowerBand = lowerBand

	return frasma2Val
}

// UpdateAll updates the indicator and returns all three outputs: frasma2, upperBand, lowerBand.
func (f *FractalBands) UpdateAll(sample float64) (float64, float64, float64) {
	frasma2 := f.Update(sample)

	f.mu.RLock()
	defer f.mu.RUnlock()

	return frasma2, f.upperBand, f.lowerBand
}

// Frasma2Value returns the last computed FRASMA2 value.
func (f *FractalBands) Frasma2Value() float64 {
	f.mu.RLock()
	defer f.mu.RUnlock()

	return f.frasma2
}

// UpperBandValue returns the last computed upper band value.
func (f *FractalBands) UpperBandValue() float64 {
	f.mu.RLock()
	defer f.mu.RUnlock()

	return f.upperBand
}

// LowerBandValue returns the last computed lower band value.
func (f *FractalBands) LowerBandValue() float64 {
	f.mu.RLock()
	defer f.mu.RUnlock()

	return f.lowerBand
}

// UpdateScalar updates the indicator given the next scalar sample.
func (f *FractalBands) UpdateScalar(sample *entities.Scalar) core.Output {
	frasma2, upper, lower := f.UpdateAll(sample.Value)

	const outputCount = 4

	output := make([]any, outputCount)
	output[0] = entities.Scalar{Time: sample.Time, Value: frasma2}
	output[1] = entities.Scalar{Time: sample.Time, Value: upper}
	output[2] = entities.Scalar{Time: sample.Time, Value: lower}

	if math.IsNaN(lower) || math.IsNaN(upper) {
		output[3] = outputs.NewEmptyBand(sample.Time)
	} else {
		output[3] = outputs.NewBand(sample.Time, lower, upper)
	}

	return output
}

// UpdateBar updates the indicator given the next bar sample.
func (f *FractalBands) UpdateBar(sample *entities.Bar) core.Output {
	v := f.barFunc(sample)

	return f.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}

// UpdateQuote updates the indicator given the next quote sample.
func (f *FractalBands) UpdateQuote(sample *entities.Quote) core.Output {
	v := f.quoteFunc(sample)

	return f.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}

// UpdateTrade updates the indicator given the next trade sample.
func (f *FractalBands) UpdateTrade(sample *entities.Trade) core.Output {
	v := f.tradeFunc(sample)

	return f.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}

File: ts/indicators/jean-philippe-poton/fractal-bands/fractal-bands.ts
import { buildMetadata } from '../../core/build-metadata';
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorOutput } from '../../core/indicator-output';
import { IndicatorIdentifier } from '../../core/indicator-identifier';
import { LineIndicator } from '../../core/line-indicator';
import { Band } from '../../core/outputs/band';
import { Scalar } from '../../../entities/scalar';
import { Bar } from '../../../entities/bar';
import { Quote } from '../../../entities/quote';
import { Trade } from '../../../entities/trade';
import { FractalBandsParams } from './params';

/** Function to calculate mnemonic of a __FractalBands__ indicator. */
export const fractalBandsMnemonic = (params: FractalBandsParams): string =>
    'fban('.concat(
        params.period.toString(), ',',
        params.normalSpeed.toString(), ',',
        params.alpha.toString(),
        componentTripleMnemonic(params.barComponent, params.quoteComponent, params.tradeComponent),
        ')');

/** Fractal Bands indicator. */
export class FractalBands extends LineIndicator {
    private window: Array<number>;
    private _closes: Array<number>;
    private _period: number;
    private _periodMinus1: number;
    private _normalSpeed: number;
    private _alpha: number;
    private windowCount: number;
    private logDenom: number;
    private ln2: number;
    private invPeriodSq: number;
    private _frasma2: number;
    private _upperBand: number;
    private _lowerBand: number;

    /**
     * Constructs an instance given parameters.
     * Period should be greater than 1, normalSpeed greater than 0, alpha greater than 0.
     **/
    public constructor(params: FractalBandsParams) {
        super();
        const period = Math.floor(params.period);
        const normalSpeed = Math.floor(params.normalSpeed);
        const alpha = params.alpha;

        if (period < 2) {
            throw new Error('period should be greater than 1');
        }

        if (normalSpeed < 1) {
            throw new Error('normal_speed should be greater than 0');
        }

        if (alpha <= 0.0) {
            throw new Error('alpha should be greater than 0');
        }

        this.mnemonic = fractalBandsMnemonic(params);
        this.description = 'Fractal bands ' + this.mnemonic;
        this.barComponent = params.barComponent;
        this.quoteComponent = params.quoteComponent;
        this.tradeComponent = params.tradeComponent;
        this.window = new Array<number>(period);
        this._closes = [];
        this._period = period;
        this._periodMinus1 = period - 1;
        this._normalSpeed = normalSpeed;
        this._alpha = alpha;
        this.windowCount = 0;
        this.primed = false;
        this.logDenom = Math.log(2.0 * (period - 1));
        this.ln2 = Math.log(2.0);
        this.invPeriodSq = 1.0 / (period * period);
        this._frasma2 = Number.NaN;
        this._upperBand = Number.NaN;
        this._lowerBand = Number.NaN;
    }

    /** Describes the output data of the indicator. */
    public metadata(): IndicatorMetadata {
        return buildMetadata(
            IndicatorIdentifier.FractalBands,
            this.mnemonic,
            this.description,
            [
                { mnemonic: this.mnemonic, description: this.description },
                { mnemonic: this.mnemonic + ' upper', description: this.description + ' Upper Band' },
                { mnemonic: this.mnemonic + ' lower', description: this.description + ' Lower Band' },
                { mnemonic: this.mnemonic + ' band', description: this.description + ' Band' },
            ],
        );
    }

    /** Updates the indicator and returns the FRASMA2 value. */
    public update(sample: number): number {
        if (Number.isNaN(sample)) {
            return sample;
        }

        const period = this._period;
        const periodMinus1 = this._periodMinus1;

        // Accumulate close history for SMA computation.
        this._closes.push(sample);

        // Fill the FGDI window.
        if (this.primed) {
            for (let i = 0; i < periodMinus1; i++) {
                this.window[i] = this.window[i + 1];
            }

            this.window[periodMinus1] = sample;
        } else {
            this.window[this.windowCount] = sample;
            this.windowCount++;

            if (this.windowCount < period) {
                return Number.NaN;
            }

            this.primed = true;
        }

        // Find min/max for normalization.
        let priceMax = this.window[0];
        let priceMin = this.window[0];

        for (let k = 1; k < period; k++) {
            if (this.window[k] > priceMax) { priceMax = this.window[k]; }
            if (this.window[k] < priceMin) { priceMin = this.window[k]; }
        }

        const priceRange = priceMax - priceMin;

        let fgdi: number;

        if (priceRange <= 0.0) {
            fgdi = 0.0;
        } else {
            // Compute normalized path length: period points, period-1 segments.
            let priorNorm = (this.window[0] - priceMin) / priceRange;
            let length = 0.0;

            for (let k = 1; k < period; k++) {
                const currNorm = (this.window[k] - priceMin) / priceRange;
                const diff = currNorm - priorNorm;
                length += Math.sqrt(diff * diff + this.invPeriodSq);
                priorNorm = currNorm;
            }

            if (length > 0.0) {
                fgdi = 1.0 + (Math.log(length) + this.ln2) / this.logDenom;
            } else {
                fgdi = 0.0;
            }
        }

        // Hurst exponent.
        let hurst = 2.0 - fgdi;
        if (hurst < 0.01) {
            hurst = 0.01;
        }

        const trailDim = 1.0 / hurst;
        const beta = trailDim / 2.0;
        let speed = Math.max(Math.round(this._normalSpeed * beta), 1);

        // FRASMA2: SMA of close over 'speed' bars ending at current position.
        const nCloses = this._closes.length;
        if (speed > nCloses) {
            this._frasma2 = Number.NaN;
            this._upperBand = Number.NaN;
            this._lowerBand = Number.NaN;
            return Number.NaN;
        }

        let smaSum = 0.0;
        for (let k = nCloses - speed; k < nCloses; k++) {
            smaSum += this._closes[k];
        }

        const frasma2Val = smaSum / speed;

        // Deviation over the FGDI lookback window (period bars).
        let sqSum = 0.0;
        for (let k = 0; k < period; k++) {
            const res = this.window[k] - frasma2Val;
            sqSum += res * res;
        }

        const deviation = 2.0 * Math.sqrt(sqSum / period);

        // Fractal bands.
        const bandMult = deviation * Math.pow(this._alpha, hurst);
        const upperBand = frasma2Val + bandMult;
        const lowerBand = frasma2Val - bandMult;

        this._frasma2 = frasma2Val;
        this._upperBand = upperBand;
        this._lowerBand = lowerBand;

        return frasma2Val;
    }

    /** Updates and returns all three outputs: [frasma2, upperBand, lowerBand]. */
    public updateAll(sample: number): [number, number, number] {
        const frasma2 = this.update(sample);
        return [frasma2, this._upperBand, this._lowerBand];
    }

    /** Returns the last computed FRASMA2 value. */
    public get frasma2Value(): number { return this._frasma2; }

    /** Returns the last computed upper band value. */
    public get upperBandValue(): number { return this._upperBand; }

    /** Returns the last computed lower band value. */
    public get lowerBandValue(): number { return this._lowerBand; }

    /** Updates the indicator given the next scalar sample. */
    public updateScalar(sample: Scalar): IndicatorOutput {
        const [frasma2, upper, lower] = this.updateAll(sample.value);

        const s0 = new Scalar(); s0.time = sample.time; s0.value = frasma2;
        const s1 = new Scalar(); s1.time = sample.time; s1.value = upper;
        const s2 = new Scalar(); s2.time = sample.time; s2.value = lower;

        const band = new Band();
        band.time = sample.time;
        if (isNaN(lower) || isNaN(upper)) {
            band.lower = NaN;
            band.upper = NaN;
        } else {
            band.lower = lower;
            band.upper = upper;
        }

        return [s0, s1, s2, band];
    }

    /** Updates the indicator given the next bar sample. */
    public updateBar(sample: Bar): IndicatorOutput {
        const v = this.barComponentFunc(sample);
        const scalar = new Scalar();
        scalar.time = sample.time;
        scalar.value = v;
        return this.updateScalar(scalar);
    }

    /** Updates the indicator given the next quote sample. */
    public updateQuote(sample: Quote): IndicatorOutput {
        const v = this.quoteComponentFunc(sample);
        const scalar = new Scalar();
        scalar.time = sample.time;
        scalar.value = v;
        return this.updateScalar(scalar);
    }

    /** Updates the indicator given the next trade sample. */
    public updateTrade(sample: Trade): IndicatorOutput {
        const v = this.tradeComponentFunc(sample);
        const scalar = new Scalar();
        scalar.time = sample.time;
        scalar.value = v;
        return this.updateScalar(scalar);
    }
}
